shared-libs/transitions/src/lib/muting_utils.js

```diff
@@ -28,26 +28,69 @@ const updateContacts = (contacts, muted) => {
     return Promise.resolve();
   }
 
-  contacts.forEach(contact => updateContact(contact, muted));
-  return db.medic.bulkDocs(contacts);
+  const updatedContacts = contacts.filter(contact => updateContact(contact, muted));
+  if (!updatedContacts.length) {
+    return Promise.resolve();
+  }
+
+  return db.medic.bulkDocs(updatedContacts);
 };
 
+/**
+ * Updates the muted state on a given contact, updates local muting history, if it exists.
+ * If a contact does not need updating, the object is not mutated and the function returns boolean false.
+ * The contact does not need updating when either:
+ * a) an already muted contact is muted and there is no muting history
+ * b) an already unmuted contact is unmuted and there is no muting history
+ * c) an already muted contact is muted, and the muting history is up to date
+ * d) an already unmuted contact is unmuted, and the muting history is up to date
+ * Muting history is up to date when:
+ * - the last update was made on the server
+ * - the server muted state corresponds with the updated muted state.
+ * In any other case, the contact will be updated with the new muted state.
+ * @param {Object} contact
+ * @param {moment|boolean} muted
+ * @return {boolean}
+ */
 const updateContact = (contact, muted) => {
+  const isMutingHistoryUpToDate = !contact.muting_history ||
+                                  (
+                                    contact.muting_history &&
+                                    contact.muting_history.last_update === SERVER &&
+                                    contact.muting_history[SERVER].muted === !!muted
+                                  );
+
+  let updated = false;
+  let mutedTimestamp;
   if (muted) {
-    contact.muted = muted;
-  } else {
-    delete contact.muted;
+    if (isMutingHistoryUpToDate) {
+      // if muting an already muted contact and muting history requires no changes, use the contact's muted timestamp
+      mutedTimestamp = contact.muted || muted;
+    } else {
+      // if muting an already muted contact and muting history requires changes, use the new muted timestamp
+      mutedTimestamp = muted;
+    }
   }
 
-  if (contact.muting_history) {
+  if (!isMutingHistoryUpToDate && contact.muting_history) {
+    updated = true;
     contact.muting_history[SERVER] = {
       muted: !!muted,
-      date: muted || new Date().getTime(),
+      date: mutedTimestamp || moment(),
```

Comment:
We don't explicitly do `toISOString` anywhere. I think that moment stringifies into the iso date by default. 
If I do: 
```
const a = { date: moment() };
console.log(JSON.stringify(a));
```
I get: `{"date":"2021-07-29T07:54:29.758Z"}`

https://momentjscom.readthedocs.io/en/latest/moment/04-displaying/13-as-json/

Reply:
> We don't explicitly do toISOString anywhere.

I did a quick search of the code and found quite a few examples of using `toISOString()`, for example, in the task history code which is very similar to this. I feel much more comfortable explicitly calling it rather than relying on the default moment behaviour because a) they may change the behaviour, and b) it's not obvious to the reader what it's going to do.

Reply:
Ok, cool, I've changed it so I don't pass a moment object, and instead pass the ISO string. Makes comparisons easier too.